pkg/metrics/metrics.go

```diff
@@ -0,0 +1,271 @@
+package metrics
+
+import (
+	"errors"
+	"fmt"
+
+	"github.com/kanisterio/kanister/pkg/log"
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+// BoundedLabel is a type that represents a label and its associated
+// valid values
+type BoundedLabel struct {
+	LabelName   string
+	LabelValues []string
+}
+
+// getLabelNames extracts the all the `LabelName` fields from each BoundedLabel struct
+func getLabelNames(bl []BoundedLabel) []string {
+	ln := make([]string, 0)
+	for _, l := range bl {
+		ln = append(ln, l.LabelName)
+	}
+	return ln
+}
+
+// generateLabelCombinations uses a backtracking approach to generate
+// a list of label permutations
+func generateLabelCombinations(bl []BoundedLabel, labelIndex int,
+	workingSlice [][]string, resultCombinations *[][][]string) (bool, error) {
```

Comment:
```suggestion
func generateLabelCombinations(
            bl []BoundedLabel, labelIndex int,
            workingSlice [][]string,
            resultCombinations *[][][]string
            ) (bool, error) {
```

Reply:
Thanks for pointing this out. I've changed it to fillLabelCombinations because we're "filling" a list reference passed as a function argument. 
Let me know if this is okay.